rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+map into LLVM's non-atomic volatile operations, unnecessarily differ from the
+memory access semantics of mainstream hardware in two major ways:
+
+1. Concurrent use of volatile memory operations on a given memory location is
+   considered to be Undefined Behavior, and may therefore result in unintended
+   compilation output if detected by the optimizer.
+2. Using an overly wide volatile load or store operation which cannot be carried
+   out by a single hardware load and store instruction will not result in a
+   compilation error, but in the silent emission of multiple hardware load or
+   store instructions, which might be a logic error in the users' program.
+
+By implementing support for LLVM's atomic volatile operations, and encouraging
+their use on every hardware that supports them, we eliminate these divergences
+from hardware behavior and therefore bring volatile operations closer to the
+"defer to hardware memory model" semantics that programmers expect them to have.
+This reduces the odd of mistake in programs operating outside of the regular
+Rust Abstract Machine semantics, which are notoriously hard to get right.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Rust compiler generally assumes that the program that it is building is
+living in a fully isolated memory space, where the contents of memory can only
+change if some direct action from the program (including FFI or atomic memory
+operations) allows it to change.
+
+It leverages this knowledge to transform said program's memory access patterns
+for performance optimization purposes, under the assumption that said
+transformations will not have any externally observable effect other than
+speeding up the program.
+
+Examples of such transformations include:
+
+- caching data from RAM into CPU registers,
+- eliminating unused loads and unobservable stores,
+- merging neighboring stores with each other,
+- only updating the part of an over-written struct that has actually changed.
+
+Although these optimizations are most of the time correct and useful, there are
+some situations where they are inappropriate, in areas such as:
+
+- [Memory-mapped I/O](https://en.wikipedia.org/wiki/Memory-mapped_I/O), a common
+  low-level communication protocol between CPUs and peripherals, where hardware
+  registers masquerading as memory can be used to program peripherals by
+  accessing said registers in very specific load/store patterns.
+- [Shared memory](https://en.wikipedia.org/wiki/Shared_memory), a form of
+  inter-process communication where two programs can communicate via a common
+  memory block, which means that stores are externally observable and loads are
+  not guaranteed to return consistent results.
+- Advanced uses of [virtual memory](https://en.wikipedia.org/wiki/Virtual_memory)
+  where the mere action of reading from or writing to memory may trigger
+  execution of arbitrary code by the operating system.
```

Comment:
Note that LLVM [explicitly limits the allowed side effects of volatile accesses](https://llvm.org/docs/LangRef.html#volatile-memory-accesses) (to retain ability to to optimize *other* code in the vicinity of volatile accesses):

> A volatile operation may not modify any other memory accessible by the module being compiled. A volatile operation may not call any code in the current module.

So depending on what you have in mind, virtual memory shenanigans may not work even with volatile accesses. For example, I recall a "pauseless" GC design (by Azul) that lazily fixes up references to relocated objects by setting up page tables to trap on reads from the old location of relocated objects. This requires updating memory locations other than the one loaded from, so it's an example of virtual memory trickery that is not sanctioned by LLVM's volatile semantics. (Of course, in practice Azul probably doesn't use volatile accesses for literally everything, so this specific example is not directly applicable anyway.)

Reply:
Thanks, I'll try to find a concise way to word this.

Reply:
Added!